### ISSUE TITLE
How is touchscreen cursor emulation supposed to work?

### ISSUE THREAD

**Issue by spl237** [labwc/labwc#2242]
In touch.c, the functions touch_down, touch_up and touch_motion can all call cursor emulation functions in cursor.c (cursor_emulate_move_absolute and cursor_emulate_button). I've been playing with this - it is currently impossible to generate a double-click on a touchscreen by tapping, as there seems to be no such thing as a double-tap equivalent to a double-click. Hence I was wondering about forcing touch events to be interpreted as mouse emulations - I found that by hacking seat_update_capabilities in seat.c so that a touch input device returned its capability as WL_SEAT_CAPABILITY_POINTER, this forced the use of the cursor emulation functions.

They don't quite work, unfortunately. If I drag a selection rectangle in the file manager, the rectangle is created on tap and resized on drag, but releasing the screen, while it does generate a cursor_emulate_button RELEASE, does not remove the selection rectangle. 

It also seems to take two taps for a button press to register on our taskbar - not sure if this is something to do with the GTK gesture recognisers I have implemented to detect tap-and-hold behaviour, but I don't think so - it seems that the "clicked" signal on a button is only generated when the button is clicked for the second time.

Is this how touch emulation is supposed to work? I note that for touchpad devices, you can set mouseEmulation to be yes in the rc.xml file, but that this has no effect for touchscreens. As a first step, if it is relatively easy to do, allowing mouseEmulation on touchscreens might be useful?

**Comment by johanmalm** [comment]
I'm not too clued up on the cursor emulation tbh. Have discussed with @jp7677 + others on IRC.

The emulation was introduced for tablets tools, but there is nothing 'tablet' specific about it.

The double-click functionality goes via `is_double_click()` in `cursor.c`

- https://github.com/labwc/labwc/blob/da68df5f5b6b1c7e8729bdfcd1e527e83f547067/src/input/cursor.c#L894

A couple have tried on touchscreens and confirms that "everything works" - i.e. single tap to open menu, double-tap on SSD titlebar to maximize window, tap-and-drag on SSD for moving a window.... except with Gtk applications! Not sure why.

**Comment by jp7677** [comment]
Trying this on a Pi touchscreen with labwc master on Raspberry OS I'm seeing the following:

gnome-terminal (GTK, SSD) 
- interaction with application, e.g. opening a menu, on tap works 
- double tap on the SSD titlebar maximizes and umaximizes
- tap-drag on the SSD titlebar moves the application window

nautilus (GTK, CSD):
- interaction with application, e.g. opening a folder, work. Also on double-tap when nautilus is configured to open folder by double click.
- double tap on a free area of the CSD titlebar maximizes and umaximizes (albeit sometimes difficult to achieve)
- tap-drag on the CSD titlebar did not move the application window, at least I didn't succeeded in doing so

The feeling of the mouse cursur isn't that natural though when interacting with an application. The cursor follows a tap/tap-move on the desktop or on an SSD titlebar, but not when interacting with an applications. That way you can end up with e.g. a hover effect on a different position than the position of the tap.

**Comment by jp7677** [comment]
> The feeling of the mouse cursur isn't that natural though when interacting with an application. The cursor follows a tap/tap-move on the desktop or on an SSD titlebar, but not when interacting with an applications. That way you can end up with e.g. a hover effect on a different position than the position of the tap.

as a test I added

```c
wlr_cursor_warp_absolute(seat->cursor, &event->touch->base, event->x, event->y);
```

before `wlr_seat_touch_notify_motion()` (https://github.com/labwc/labwc/blob/master/src/input/touch.c#L69) and `wlr_seat_touch_notify_down()` (https://github.com/labwc/labwc/blob/master/src/input/touch.c#L125), this made the cursor behavior consistent with the touch events. That said I'm not sure what the best UI is for touch. May be the cursor should follow touch input but be invisible until another input device is used?

**Comment by tokyo4j** [comment]
> They don't quite work, unfortunately. If I drag a selection rectangle in the file manager, the rectangle is created on tap and resized on drag, but releasing the screen, while it does generate a cursor_emulate_button RELEASE, does not remove the selection rectangle.

> It also seems to take two taps for a button press to register on our taskbar - not sure if this is something to do with the GTK gesture recognisers I have implemented to detect tap-and-hold behaviour, but I don't think so - it seems that the "clicked" signal on a button is only generated when the button is clicked for the second time.

This turned out to be due to missing `wl_pointer.frame` event. Wlroots emits `wlr_cursor.events.frame` event after `wlr_cursor.events.button` events (for libinput backend), but we are not emulating this behavior and only sending `wl_pointer::button` events without following `wl_pointer::frame` events.

**Comment by jp7677** [comment]
> This turned out to be due to missing `wl_pointer.frame` event. Wlroots emits `wlr_cursor.events.frame` event after `wlr_cursor.events.button` events (for libinput backend), but we are not emulating this behavior and only sending `wl_pointer::button` events without following `wl_pointer::frame` events.

ah, nice find! This might having to click twice under certain circumstance (that I had very very rarely) when using mouse emulation for tablets, which I actually use all the time.

**Comment by spl237** [comment]
It would be nice to make Hiroaki's first change above - the one in cursor.c which forces all touch events to be treated as emulated cursor events - to be a configuration option somewhere. It makes a few other things work properly; not just double-tap in GTK applications, but also some gesture recognizers in menus, but I'm not sure if it might have other side-effects, so it would be useful for the user to be able to turn it off if needed. Could we add something to rc.xml or similar for alwaysEmulateMouseWithCursor, or similar - or even have it linked to the mouseEmulation setting for a device?

**Comment by spl237** [comment]
Further to that, for now I have added a new touchscreen section to rc.xml with the one entry emulateMouse - if you set this to true then seat_update_capabilities sets the capability for a touch input device to WL_SEAT_CAPABILITY_POINTER; otherwise it leaves it as WL_SEAT_CAPABILITY_TOUCH.

This is probably not the right long-term solution, but it fixes a few things for me on Pi, so I'll include it in our forthcoming release, with emulateMouse set true by default.

**Comment by jp7677** [comment]
May be we can do the same for touch as for tablets in the long term. There we have https://labwc.github.io/labwc-config.5.html#entry_tablet_mouseemulation= which is completely handled in  `tablet.c` (starting with https://github.com/labwc/labwc/blob/master/src/input/tablet.c#L267). 

Adding the missing `wlr_seat_pointer_notify_frame(seat->seat);` is imho even be something for 0.8.1.

**Comment by spl237** [comment]
Yes, it would be consistent to have the same mouseEmulation option for touchscreens as for tablets.
